python/ray/dag/compiled_dag_node.py

```diff
@@ -1075,6 +1124,221 @@ def _get_or_compile(
+                if self.operation.idx == other.operation.idx:
+                    return self.operation.type == other.operation.type
+                return False
+
+            def __hash__(self):
+                return hash((self.operation, self.idx))
+
+        def _add_edge(in_node: GraphNode, out_node: GraphNode):
+            in_node.out_edges.add((out_node.idx, out_node.operation.type))
+            out_node.in_edges.add((in_node.idx, in_node.operation.type))
+
+        graph: Dict[int, Dict[DAGNodeOperationType, GraphNode]] = {}
+
+        from ray.dag import (
+            ClassMethodNode,
+            MultiOutputNode,
+        )
+
+        for _, executable_tasks in self.actor_to_executable_tasks.items():
+            prev_compute_node = None
+            for local_idx, exec_task in enumerate(executable_tasks):
+                idx = exec_task.idx
+                read_node = GraphNode(
+                    DAGNodeOperation(local_idx, DAGNodeOperationType.READ),
+                    idx,
+                    self.idx_to_task[idx].dag_node,
+                )
+                compute_node = GraphNode(
+                    DAGNodeOperation(local_idx, DAGNodeOperationType.COMPUTE),
+                    idx,
+                    self.idx_to_task[idx].dag_node,
+                )
+                write_node = GraphNode(
+                    DAGNodeOperation(local_idx, DAGNodeOperationType.WRITE),
+                    idx,
+                    self.idx_to_task[idx].dag_node,
+                )
+                _add_edge(read_node, compute_node)
+                _add_edge(compute_node, write_node)
+                if prev_compute_node is not None:
+                    _add_edge(prev_compute_node, compute_node)
+                prev_compute_node = compute_node
+                graph[idx] = {
+                    DAGNodeOperationType.READ: read_node,
+                    DAGNodeOperationType.COMPUTE: compute_node,
+                    DAGNodeOperationType.WRITE: write_node,
+                }
+
+        for idx, task in self.idx_to_task.items():
+            if not isinstance(task.dag_node, ClassMethodNode):
+                continue
+            for downstream_idx in task.downstream_node_idxs:
+                downstream_dag_node = self.idx_to_task[downstream_idx].dag_node
+                if isinstance(downstream_dag_node, MultiOutputNode):
+                    continue
+                _add_edge(
+                    graph[idx][DAGNodeOperationType.WRITE],
+                    graph[downstream_idx][DAGNodeOperationType.READ],
+                )
+
+        import heapq
+
+        actor_to_candidates = defaultdict(list)
+        for idx, node_dict in graph.items():
+            for _, node in node_dict.items():
+                if node.in_degree == 0:
+                    heapq.heappush(actor_to_candidates[node.actor_handle], node)
+
+        visited_nodes = set()
+
+        def _select_next_nodes():
```

Comment:
Make it a top-level function.

Reply:
Updated https://github.com/ray-project/ray/pull/46911/commits/bf6957060d2209d1bd9c827f56b5f45524f13541